README.md
# feathers-http-distributed

[![Build Status](https://travis-ci.org/dekelev/feathers-http-distributed.svg?branch=master)](https://travis-ci.org/dekelev/feathers-http-distributed)
[![Coverage Status](https://coveralls.io/repos/github/dekelev/feathers-http-distributed/badge.svg?branch=master)](https://coveralls.io/github/dekelev/feathers-http-distributed?branch=master)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat-square)](https://github.com/standard/semistandard)
[![Dependency Status](https://img.shields.io/david/dekelev/feathers-http-distributed.svg)](https://david-dm.org/dekelev/feathers-http-distributed)
[![npm](https://img.shields.io/npm/v/feathers-http-distributed.svg?maxAge=3600)](https://www.npmjs.com/package/feathers-http-distributed)

Distribute FeathersJS apps over the network with inter-service communication using HTTP protocol.

### Install

```bash
npm install --save feathers-http-distributed
```

### Init

```js
const feathers = require('@feathersjs/feathers');
const distributed = require('feathers-http-distributed');

const app = feathers();

app.configure(distributed({}));
```

##### Options

| Option | Type | Default | Required | Description |
| --- |:---: | :---: | :---: | --- |
| protocol | string | `'http'` | no | Protocol to use when calling remote services.<br/><br/>Supported protocols are `http` & `https`. |
| host | string | `null` | no | Default hostname to use when calling remote services. |
| port | number | `80` | no | Port number to use when calling remote services.<br/><br/>Defaults to the default port of the selected protocol. |
| dnsSuffix | string | `''` | no | DNS suffix that will be added to the `host` when calling remote services. |
| pathToHost | boolean<br/>function | `false` | no | If `host` is not set, path will be converted into `host` by replacing all the non-alphanumeric characters to `-`.<br/><br/>Can also be set with a custom method that receives a path and returns a `host`. |
| timeout | number | `0` | no | Request timeout in milliseconds.<br/><br/>Set to `0` to disable timeout.<br/><br/>If timeout is enabled, by default, it will include the time spent on retries. |
| proxy | object | `null` | no | Transparent HTTP proxy to forward requests to remote services.<br/><br/>Set the `proxy` object with `host` & `port`.<br/><br/>If proxy authentication is required, set the `auth` key with object containing the `username` & `password` keys. |
| excludeParams | string[] | `null` | no | List of keys to exclude from the `params` object of the remote service call before sending the request. |
| maxRedirects | number | `5` | no | Maximum redirects to follow.<br/><br/>Set to `0` to disable redirects. |
| keepAlive | boolean | `false` | no | Use HTTP persistent connections with HTTP keep-alive. |
| internalRequestHeader | string | `X-Internal-Request` | no | Name of the request header that is sent with each request to remote service.<br/><br/>This header is used to identify the request as internal and contains the `params` object of the service call.<br/><br/>**Add rule in your external API Gateway or load-balancer to remove this header from all incoming requests.** |
| retry | boolean<br/>object | `false` | no | Retry failed requests on a network error or when receiving 5xx error on an idempotent request (GET, HEAD, OPTIONS, PUT or DELETE).<br/><br/>By default, it will retry failed requests 3 times without delay.<br/><br/>List of all the supported retry options is available [here](https://www.npmjs.com/package/axios-retry#options). |

### Call remote service

```js
const result = await app.service('remote').find({});

const result = await app.service('remote').find({ host: 'remote-app' });
```

##### Params

| Option | Type | Required | Description |
| --- | :---: | :---: | --- |
| protocol | string | no | Overrides the `protocol` init option. |
| host | string | no | Overrides the `host` and `pathToHost` init options. |
| port | number | no | Overrides the `port` init option. |
| dnsSuffix | string | no | Overrides the `dnsSuffix` init option. |
| timeout | number | no | Overrides the `timeout` init option. |
| proxy | object | no | Overrides the `proxy` init option. |

### Middleware

Use the `handleInternalRequest` method to detect and handle incoming HTTP requests from remote FeathersJS apps.
  
When `handleInternalRequest` returns `true`, skip any further custom middlewares that should only apply to external HTTP requests.

```js
const { handleInternalRequest } = require('feathers-http-distributed');

app.use((req, res, next) => {
  if (handleInternalRequest(req)) {
    next();

    return;
  }

  // Add here custom middlewares that only applies to external HTTP requests
});
```

### Security

Secure your network by adding a rule in your external API Gateway or load-balancer to remove the `X-Internal-Request` request header from all the incoming requests.

### Debug logs

Debug logs can be enabled by settings the `DEBUG` environment variable to `feathers-http-distributed*,axios`.

### Debugging in Kubernetes

The `proxy` option can be used to forward requests from a FeathersJS app running locally to remote services inside Kubernetes clusters with the help of transparent HTTP proxies.

Tools like [Telepresence](https://www.telepresence.io/) helps with debugging incoming traffic that goes into Kubernetes pods, by swapping the pods with proxy pods and redirects incoming traffic to a local port on the host.

With the `proxy` option set, you can simply run [Telepresence](https://www.telepresence.io/) with the [`inject-tcp` proxying method](https://www.telepresence.io/reference/methods.html) and debug your FeathersJS app as you normally do.

See [here](https://github.com/dekelev/kong-transparent-proxy) for example of deploying transparent HTTP proxy with Docker.


CHANGELOG.md
# Changelog

## [v1.1.0](https://github.com/dekelev/feathers-http-distributed/tree/v1.1.0) (2023-03-04)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v1.0.4...v1.1.0)

## [v1.0.4](https://github.com/dekelev/feathers-http-distributed/tree/v1.0.4) (2022-01-15)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v1.0.3...v1.0.4)

## [v1.0.3](https://github.com/dekelev/feathers-http-distributed/tree/v1.0.3) (2021-07-10)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v1.0.2...v1.0.3)

**Merged pull requests:**

- Bump glob-parent from 5.1.1 to 5.1.2 [\#6](https://github.com/dekelev/feathers-http-distributed/pull/6) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump hosted-git-info from 2.8.8 to 2.8.9 [\#5](https://github.com/dekelev/feathers-http-distributed/pull/5) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump lodash from 4.17.20 to 4.17.21 [\#4](https://github.com/dekelev/feathers-http-distributed/pull/4) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v1.0.2](https://github.com/dekelev/feathers-http-distributed/tree/v1.0.2) (2021-03-30)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v1.0.1...v1.0.2)

**Merged pull requests:**

- Bump y18n from 4.0.0 to 4.0.1 [\#3](https://github.com/dekelev/feathers-http-distributed/pull/3) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v1.0.1](https://github.com/dekelev/feathers-http-distributed/tree/v1.0.1) (2021-03-30)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v1.0.0...v1.0.1)

**Closed issues:**

- Upgrade axios for high severity vulnerability [\#1](https://github.com/dekelev/feathers-http-distributed/issues/1)

## [v1.0.0](https://github.com/dekelev/feathers-http-distributed/tree/v1.0.0) (2021-03-02)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v0.9.3...v1.0.0)

## [v0.9.3](https://github.com/dekelev/feathers-http-distributed/tree/v0.9.3) (2020-08-17)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v0.9.2...v0.9.3)

## [v0.9.2](https://github.com/dekelev/feathers-http-distributed/tree/v0.9.2) (2020-08-17)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v0.9.1...v0.9.2)

## [v0.9.1](https://github.com/dekelev/feathers-http-distributed/tree/v0.9.1) (2020-08-17)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/v0.9.0...v0.9.1)

## [v0.9.0](https://github.com/dekelev/feathers-http-distributed/tree/v0.9.0) (2020-08-17)

[Full Changelog](https://github.com/dekelev/feathers-http-distributed/compare/c0a4b887fc5a55e998fcae19b42836e35a8331b5...v0.9.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


lib/utils.js
const { INTERNAL_REQUEST_HEADER } = require('./constants');

// Handle internal network requests from remote apps
// Returns true when internal network request is received
// Any custom middlewares for external requests should be skipped when receiving an internal network request
const handleInternalRequest = function (req, options = {}) {
  const internalRequestHeader = options.internalRequestHeader || INTERNAL_REQUEST_HEADER;
  const requestHeader = req.headers[internalRequestHeader.toLowerCase()];

  if (!requestHeader) { return false; }

  const params = JSON.parse(requestHeader);

  params.fromRemote = true;

  for (const key of Object.keys(params)) { req.feathers[key] = params[key]; }

  delete req.feathers.provider;
  delete req.feathers.headers;

  return true;
};

// Converts path to hostname
const pathToHost = function (path) {
  return path.replace(/[^a-z0-9]/gi, '-');
};

// Converts Feathers id to string
const idToString = function (id) {
  if (typeof id === 'object') { return JSON.stringify(id); }

  return id;
};

// Get the default port of protocol
const getProtocolPort = function (protocol) {
  if (protocol === 'http') { return 80; } else if (protocol === 'https') { return 443; }
};

module.exports = {
  handleInternalRequest,
  pathToHost,
  idToString,
  getProtocolPort
};


lib/service.js
const makeDebug = require('debug');
const errorHandler = require('./error-handler');

const debug = makeDebug('feathers-http-distributed:service');

class RemoteService {
  constructor (path, requester) {
    this.path = path;
    this.requester = requester;
    this.remote = true;
  }

  async find (params) {
    debug('Requesting find() remote service on path ' + this.path, this.requester.filterParams(params));
    try {
      const result = await this.requester.send({ type: 'find', path: this.path, params });
      debug('Successfully find() remote service on path ' + this.path);
      return result;
    } catch (error) {
      throw errorHandler(error);
    }
  }

  async get (id, params) {
    debug('Requesting get() remote service on path ' + this.path, id, this.requester.filterParams(params));
    try {
      const result = await this.requester.send({ type: 'get', path: this.path, id, params });
      debug('Successfully get() remote service on path ' + this.path);
      return result;
    } catch (error) {
      throw errorHandler(error);
    }
  }

  async create (data, params) {
    debug('Requesting create() remote service on path ' + this.path, data, this.requester.filterParams(params));
    try {
      const result = await this.requester.send({ type: 'create', path: this.path, data, params });
      debug('Successfully create() remote service on path ' + this.path);
      return result;
    } catch (error) {
      throw errorHandler(error);
    }
  }

  async update (id, data, params) {
    debug('Requesting update() remote service on path ' + this.path, id, data, this.requester.filterParams(params));
    try {
      const result = await this.requester.send({ type: 'update', path: this.path, id, data, params });
      debug('Successfully update() remote service on path ' + this.path);
      return result;
    } catch (error) {
      throw errorHandler(error);
    }
  }

  async patch (id, data, params) {
    debug('Requesting patch() remote service on path ' + this.path, id, data, this.requester.filterParams(params));
    try {
      const result = await this.requester.send({ type: 'patch', path: this.path, id, data, params });
      debug('Successfully patch() remote service on path ' + this.path);
      return result;
    } catch (error) {
      throw errorHandler(error);
    }
  }

  async remove (id, params) {
    debug('Requesting remove() remote service on path ' + this.path, id, this.requester.filterParams(params));
    try {
      const result = await this.requester.send({ type: 'remove', path: this.path, id, params });
      debug('Successfully remove() remote service on path ' + this.path);
      return result;
    } catch (error) {
      throw errorHandler(error);
    }
  }
}

module.exports = RemoteService;


lib/middleware.js
const { INTERNAL_REQUEST_HEADER } = require('./constants');

// Handle internal network requests from remote apps
// Returns true when internal network request is received
// Any custom middlewares for external requests should be skipped when receiving an internal network request
const handleInternalRequest = function (req, options = {}) {
  const internalRequestHeader = options.internalRequestHeader || INTERNAL_REQUEST_HEADER;
  const requestHeader = req.headers[internalRequestHeader.toLowerCase()];

  if (!requestHeader) { return false; }

  const params = JSON.parse(requestHeader);

  params.fromRemote = true;

  for (const key of Object.keys(params)) { req.feathers[key] = params[key]; }

  delete req.feathers.provider;
  delete req.feathers.headers;

  return true;
};

module.exports = handleInternalRequest;


lib/requester.js
const http = require('http');
const https = require('https');
const axios = require('axios');
const axiosRetry = require('axios-retry');
const errors = require('@feathersjs/errors');
const {
  DEFAULT_PROTOCOL,
  DEFAULT_TIMEOUT,
  INTERNAL_REQUEST_HEADER,
  AXIOS_HTTP_METHODS
} = require('./constants');

class Requester {
  constructor (options) {
    this.protocol = options.protocol || DEFAULT_PROTOCOL;
    this.host = options.host;
    this.port = options.port;
    this.dnsSuffix = options.dnsSuffix || '';
    this.pathToHost = options.pathToHost === true ? this.getHostByPath : options.pathToHost;
    this.timeout = options.timeout || DEFAULT_TIMEOUT;
    this.proxy = options.proxy;
    this.excludeParams = options.excludeParams;
    this.maxRedirects = options.maxRedirects;
    this.keepAlive = options.keepAlive;
    this.internalRequestHeader = options.internalRequestHeader || INTERNAL_REQUEST_HEADER;

    if (options.retry) { axiosRetry(axios, options.retry); }
  }

  async send (options) {
    const { type, path, id, data, params } = options;

    const serviceProtocol = params.protocol || this.protocol;
    const serviceHostPrefix = (params.host || this.host || (this.pathToHost && this.pathToHost(path)));
    const servicePort = params.port || this.port || this.getProtocolPort(serviceProtocol);

    this.validateProtocol(serviceProtocol);
    this.validateHost(serviceHostPrefix);
    this.validatePort(servicePort);

    const dnsSuffix = params.dnsSuffix !== undefined ? params.dnsSuffix : this.dnsSuffix;
    const serviceHost = serviceHostPrefix + dnsSuffix;

    const url = this.getUrl(serviceProtocol, serviceHost, servicePort, path, id);

    let proxy = {
      ...this.proxy,
      ...params.proxy
    };

    if (!Object.keys(proxy).length) { proxy = false; }

    const requestOptions = {
      proxy,
      headers: {
        'Content-Type': 'application/json',
        [this.internalRequestHeader]: JSON.stringify(this.filterParams(params)),
        ...params.headers
      },
      timeout: params.timeout !== undefined ? params.timeout : this.timeout
    };

    if (this.maxRedirects !== undefined) { requestOptions.maxRedirects = this.maxRedirects; }

    if (this.keepAlive) {
      requestOptions.httpAgent = new http.Agent({ keepAlive: true });
      requestOptions.httpsAgent = new https.Agent({ keepAlive: true });
    }

    const httpMethod = AXIOS_HTTP_METHODS[type];
    const args = data ? [url, data, requestOptions] : [url, requestOptions];
    const result = await axios[httpMethod](...args);

    return result.data;
  }

  validateProtocol (value) {
    if (value !== 'http' && value !== 'https') { throw new errors.BadRequest(`Invalid protocol ${value}`); }
  }

  validateHost (value) {
    if (!value) { throw new errors.BadRequest('Missing host'); }
  }

  validatePort (value) {
    if (!(value > 0 && value <= 65535)) { throw new errors.BadRequest(`Invalid port ${value}`); }
  }

  getProtocolPort (protocol) {
    if (protocol === 'http') { return 80; } else if (protocol === 'https') { return 443; }
  }

  getHostByPath (path) {
    return path.replace(/[^a-z0-9]/gi, '-');
  }

  getUrl (protocol, host, port, path, id) {
    const fullPath = id ? `${path}/${this.idToString(id)}` : path;
    const isKnownPort = (protocol === 'http' && port === 80) || (protocol === 'https' && port === 443);
    let url = `${protocol}://${host}`;

    if (!isKnownPort) { url += `:${port}`; }

    url += `/${fullPath}`;

    return url;
  }

  filterParams (params) {
    if (!this.excludeParams) { return params; }

    const result = { ...params };

    for (const param of this.excludeParams) { delete result[param]; }

    return result;
  }

  idToString (id) {
    if (typeof id === 'object') { return JSON.stringify(id); }

    return id;
  }
}

module.exports = Requester;
